bunner-macroquad/src/grass.rs

```diff
@@ -0,0 +1,201 @@
+                Self::first_hedge_row(index)
+            } else if previous_hedge_row == HedgeRow::First {
+                (previous_hedge_mask.unwrap(), HedgeRow::Second)
+            } else {
+                (Vec::new(), HedgeRow::None)
+            };
+
+        let mut children: Vec<Child> = Vec::new();
+        if hedge_row != HedgeRow::None {
+            // See comments in classify_hedge_segment for explanation of previous_mid_segment
+            #[allow(unused_assignments)]
+            let mut hedge_tile = HedgeTile::Grass;
+            let mut previous_mid_segment = None;
+            for i in 1..13 {
+                (hedge_tile, previous_mid_segment) =
+                    Self::classify_hedge_segment(&hedge_mask[i - 1..i + 2], previous_mid_segment);
+                if hedge_tile != HedgeTile::Grass {
+                    children.push(Child::Hedge(Hedge::new(
+                        hedge_tile,
+                        hedge_row,
+                        Position::new(i as i32 * 40 - 20, 0),
+                    )));
+                }
+            }
+        }
+
+        Self {
+            y,
+            index,
+            hedge_row,
+            hedge_mask,
+            children,
+        }
+    }
+
+    pub fn without_hedge(index: i32, y: i32) -> Self {
+        Self::new(None, HedgeRow::None, index, y)
+    }
+
+    pub fn classify_hedge_segment(
+        mask_window: &[HedgeMask],
+        previous_mid_segment: Option<HedgeTile>,
+    ) -> (HedgeTile, Option<HedgeTile>) {
+        if mask_window[1] == HedgeMask::Empty {
+            (HedgeTile::Grass, None)
+        } else if mask_window[0] == HedgeMask::Empty && mask_window[2] == HedgeMask::Empty {
+            (HedgeTile::SingleWidth, None)
+        } else if mask_window[0] == HedgeMask::Empty {
+            (HedgeTile::LeftMost, None)
+        } else if mask_window[2] == HedgeMask::Empty {
+            (HedgeTile::RightMost, None)
+        } else {
+            match previous_mid_segment {
+                Some(HedgeTile::Middle4) if mask_window[2] == HedgeMask::Empty => {
+                    (HedgeTile::Middle5, None)
+                }
+                Some(HedgeTile::Middle4) if mask_window[2] == HedgeMask::Hedge => {
+                    (HedgeTile::Middle3, Some(HedgeTile::Middle3))
+                }
+                Some(HedgeTile::Middle3) => (HedgeTile::Middle3, Some(HedgeTile::Middle3)),
+                _ => (HedgeTile::Middle3, Some(HedgeTile::Middle3)),
+            }
+        }
+    }
+
+    pub fn first_hedge_row(index: i32) -> (Vec<HedgeMask>, HedgeRow) {
+        if rand::gen_range::<u8>(0, 1) == 0 && index > 7 && index < 14 {
+            (Self::generate_hedge_mask(), HedgeRow::First)
+        } else {
+            (Vec::new(), HedgeRow::None)
+        }
+    }
+
+    pub fn generate_hedge_mask() -> Vec<HedgeMask> {
+        let mut mask = Vec::new();
+        mask.resize_with(12, || {
+            if rand::gen_range::<u8>(1, 100) > 1 {
```

Comment:
Super tiny difference :grimacing: `gen_range(1,100) > 1` yields a range of [1, 99] (both ends included), so the probability is 98/99, instead of 99/100 :smile: In case 99% is intended, should be `gen_range::<u8>(0, 100) >= 1`, however, this doesn't impact the game, so it's ok for me :smile:

Reply:
I thought I fixed all of these. Good catch!